agent/consul/leader.go

```diff
@@ -1529,3 +1529,68 @@ func (s *Server) reapTombstones(index uint64) {
 		)
 	}
 }
+
+func (s *Server) DatacenterSupportsFederationStates() bool {
+	if atomic.LoadInt32(&s.dcSupportsFederationStates) != 0 {
+		return true
+	}
+
+	state := serversFederationStatesInfo{
+		supported: true,
+		found:     false,
+	}
+
+	// check if they are supported in the primary dc
+	if s.config.PrimaryDatacenter != s.config.Datacenter {
+		s.router.CheckServers(s.config.PrimaryDatacenter, state.update)
+
+		if !state.supported || !state.found {
+			s.logger.Debug("federation states are not enabled in the primary dc")
+			return false
+		}
+	}
+
+	// check the servers in the local DC
+	s.router.CheckServers(s.config.Datacenter, state.update)
+
+	if state.supported && state.found {
+		atomic.StoreInt32(&s.dcSupportsFederationStates, 1)
+		return true
+	}
+
+	s.logger.Debug("federation states are not enabled in this datacenter", "datacenter", s.config.Datacenter)
+	return false
+}
+
+type serversFederationStatesInfo struct {
+	// supported indicates whether every processed server supports federation states
+	supported bool
+
+	// found indicates that at least one server was processed
+	found bool
+}
+
+var (
+	minFederationStatesVersion = version.Must(version.NewVersion("1.8.0"))
+)
+
+func (s *serversFederationStatesInfo) update(srv *metadata.Server) bool {
+	if srv.Status != serf.StatusAlive && srv.Status != serf.StatusFailed {
+		// they are left or something so regardless we treat these servers as meeting
+		// the version requirement
+		return true
+	}
+
+	// mark that we processed at least one server
+	s.found = true
+
+	if srv.Build.GreaterThanOrEqual(minFederationStatesVersion) {
+		return true
+	}
```

Comment:
I guess we don't have any way for these servers to publish their capabilities? Version checks can sometimes be a bit brittle for this type of thing. Allowing a server to explicitly say they support a specific feature would make this quite a bit more future-proof. If we ever wanted to allow some features to be turned off a capability check would continue to work, where as a version check would fail.

This `metadata.Server` struct has a `FeatureFlags` field. Could that be used? Or could we add a similar one for capabilities?

Reply:
It's tricky. The advertisement of features and build info is done with serf tags. Serf messages have to fit into single UDP packets so there's an upper limit on how much data you can put in there. The Network Segment feature can cause `O(num_segments)` growth in the metadata today. This means that for folks using the current maximum number of segments any time we increase the metadata going into those packets we effectively reduce the number of segments they can use which is not great.

If something changed recently (@i0rek ?) to make this not as bad then I can certainly switch.

Reply:
That is quite a severe restriction, good to know. I wonder if capabilities could be done with a single bit per capability to keep the size small (assuming the encoding doesn't add a lot more for each key). I'll have to read more about serf to understand how these are encoded.

Reply:
@mkeeler reminded me that segments stopped using this mechanism in consul enterprise 1.7.3 so we don't have to worry as much about that. I'll update the PR to do it using feature flags instead of build numbers.